### PR TITLE
Fix #11677 withStatus does not check status

### LINKED ISSUE #11677: withStatus does not check if supplied status is valid
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.5.11.
* Platform and Target: nginx/1.10.3 (Ubuntu), mysql  Ver 15.1 Distrib 10.0.33-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2, FriendsOfCake/Crud 5.3

### What you did
In any controller using Crud and methods mapped to the ViewAction, you can do `$this->Crud->execute()` and all is fine and well in the world. Place a `return` in front of the statement, and you get 502 Bad Gateway errors.
```
$this->Crud->mapAction('someWorkingMethod', 'Crud.View');
$this->Crud->mapAction('someBrokenMethod', 'Crud.View');

//working
public function someWorkingMethod() {
    $this->Crud->execute();
}

// crashing
public function someBrokenMethod() {
    return $this->Crud->execute();
}
```

### What happened
Through some digging, we found the statuscode in the response is null, gets transformed to 0 and nginx loses it. The statuscode being null originates from the ViewAction missing an index `api` in its defaultConfig. We encountered this problem after we updated to Crud 5.3, which now uses `$response->withStatus()` instead of `$response->statusCode()`. `withStatus()` does not check the provided status, which `statusCode` did.

### What you expected to happen
We expected our methods to simply return the response, instead of completely crashing. The statuscode should never become null.

### PR DESCRIPTION
**Pull request by tuna-rootnet**
Add same checks as statusCode had to withStatus

Refs https://github.com/cakephp/cakephp/issues/11677

### CODE REVIEW

File: src/Http/Response.php [hunk 1]
```diff
@@ -925,6 +925,12 @@ public function getStatusCode()
      */
     public function withStatus($code, $reasonPhrase = '')
     {
+        if ($code === null) {
+            return $this->_status;
```
Comment (dereuromark) [inline src/Http/Response.php]: This read part and $code being NULL is invalid.

File: src/Http/Response.php [hunk 2]
```diff
@@ -925,6 +925,12 @@ public function getStatusCode()
...
+        }
+        if (!isset($this->_statusCodes[$code])) {
+            throw new InvalidArgumentException('Unknown status code');
```
Comment (dereuromark) [inline src/Http/Response.php]: Why is this necessary? It seems wrong the code that you can set any custom code, in the case not found it will just not set any reason phrase.

**Comment by dereuromark** [comment]
If you want to validate your status code, you should do that on app level beforehand.
I think those changes here are not good :-1:

File: src/Http/Response.php (hunk 2 above)
Comment (tuna-rootnet) [reply src/Http/Response.php]: I added it because the implementation of statusCode() explicitly checks if the provided code is known to Cake, and I image there are people relying on always having a valid response code being returned. To me, it doesn't matter if a reason phrase can be added or not, just that the responsecode is valid. But, since Cake already has this list of codes, it seemed like a good start.

File: src/Http/Response.php (hunk 1 above)
Comment (tuna-rootnet) [reply src/Http/Response.php]: You're right; statusCode used it to set the default status code to the response. Changed it to set the default response code to $new.

File: src/Http/Response.php [hunk 3]
```diff
@@ -925,8 +925,11 @@ public function getStatusCode()
      */
     public function withStatus($code, $reasonPhrase = '')
     {
+        if (!isset($this->_statusCodes[$code])) {
+            throw new InvalidArgumentException('Unknown status code');
+        }
         $new = clone $this;
-        $new->_status = $code;
+        $new->_status = $code === null ? $this->_status : $code;
```
Comment (ADmad) [inline src/Http/Response.php]: Due to the check and exception added above, the `$code === null` case will never occur, hence this check is not needed.

Comment (tuna-rootnet) [reply src/Http/Response.php]: @ADmad Haven't had my coffee yet... Removed the check, the check earlier in the method should suffice.

File: src/Http/Response.php (hunk 2 above)
Comment (dereuromark) [reply src/Http/Response.php]: but what if an application uses custom codes? will that be a Problem then?

Comment (tuna-rootnet) [reply src/Http/Response.php]: It will be, just as it was before withStatus() was available. I have no problem with applications using their own statuscodes, but since CakePHP explicitly links to the RFC's for status codes and has its own internal list of valid codes and reason phrases, I'd assume it uses them as well.

I'm not against having people chose their own status code, but it should be a valid code; be that anyting between 100 and 599 or the list in Cake itself doesn't matter to me.

File: src/Http/Response.php [hunk 4]
```diff
@@ -925,6 +925,9 @@ public function getStatusCode()
      */
     public function withStatus($code, $reasonPhrase = '')
     {
+        if (!isset($this->_statusCodes[$code])) {
+            throw new InvalidArgumentException('Unknown status code');
```
Comment (ADmad) [inline src/Http/Response.php]: You need to add `use InvalidArgumentException` at top of file.

**Comment by ADmad** [comment]
@dereuromark The [ResponseInterface](https://github.com/php-fig/http-message/blob/master/src/ResponseInterface.php#L50) has `@throws \InvalidArgumentException For invalid status code arguments.`, so I don't see any problem in having the method check and throw an exception.

**Comment by dereuromark** [comment]
All right.

File: src/Http/Response.php (hunk 4 above)
Comment (tuna-rootnet) [reply src/Http/Response.php]: @ADmad It's already present in the file at rule 27.

Comment (ADmad) [reply src/Http/Response.php]: @tuna-rootnet Sorry, I missed that.

Comment (markstory) [reply src/Http/Response.php]: Can we the user what the invalid value was? And perhaps what the acceptable range is?

**Comment by ADmad** [comment]
While I like the change I don't think throwing exception should be added in a bugfix release. Better to target this to 3.next.
